### PR TITLE
Add server-side scrolling demo

### PR DESCRIPTION
**Pull request by dinvlad**
This demo combines scrolling and server-side paging
to enable "infinite" scrolling with results fetched from
the server as the user scrolls down the list.

We always fetch the next page such that it is obvious
from the vertical scrollbar when there are more results.
This approach also helps us avoid explicit pixel calculation logic,
instead relying only on (page) events.

Once loaded, the results are never removed from the cache,
which increases memory consumption over time.
This is a minor hindrance, since in a typical scenario
(e.g. generation of the list in response to a "search" request),
we expect the most relevant results to appear at the top,
with progressively more "stale" results at the bottom,
so that a typical user won't scroll too far (e.g. beyond 100-200 results).
It is also well-known that such search queries consume
progressively more server resources and are not efficient
beyond e.g. 1000 results.

**What kind of change does this PR introduce?** (check one with "x")
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Demo

**What is the current behavior?** (You can also link to an open issue here)
There are separate demos for virtual scrolling (when all of the results are fetched in advance)
and server-side paging (without vertical scrolling).

**What is the new behavior?**
We combine these together to enable "infinite" server-side scrolling.

**Does this PR introduce a breaking change?** (check one with "x")
- [ ] Yes
- [x] No

**Other information**:
This PR was requested by @amcdnl in https://github.com/swimlane/ngx-datatable/issues/611#issuecomment-299452014

### CODE REVIEW

**Comment by amcdnl** [comment]
Awesome!

How is this different from this example? https://github.com/swimlane/ngx-datatable/blob/master/demo/paging/paging-virtual.component.ts

**Comment by dinvlad** [comment]
I noticed that example (was it added only recently though?), but to the users it appears to load everything in advance (judging by the full 100 count) - even though the code paging service does return pages in pieces, it still assigns `totalElements` which normally we would not have access too. My example queries the server really only for a small chunk in the beginning (so we won't know how many total elements there are in advance), and re-querying for the next block only if needed, when the user scrolls further (so it never uses `totalElements`).

Also, I think there's a problem with that example in
```
rows.splice(start, 0, ...pagedData.data);
```
which adds more pages on every (page), so the list grows way beyond 100 elements over time. I think it should be
```
rows.splice(start, pagedData.data.length, ...pagedData.data);
```
to achieve the effect of re-querying each chunk to avoid "stale" data (which the 'scrolling' example doesn't on purpose).

So in my version we don't have access to all of the potential elements in advance, rather we load and cache as necessary on demand, and don't display "pages" to the user at all. The net effect is the appearance of "infinite" scroll as opposed to paging (which virtual or not appears the same to the user, unless we remove 'count').

**Comment by amcdnl** [comment]
Yes, it was added recently. The fetch and then pluck was only out of demo purposes. Sounds like your demo is more complete so we will use that one instead :)

**Comment by dinvlad** [comment]
Thanks! Let me know if you need anything else.

**Comment by BBaysinger** [comment]
A huge thanks for this. And just in time. I almost didn't find it as the demo site isn't updated with this yet.

**Comment by dinvlad** [comment]
Actually, I'm not completely satisfied with this solution either, since it requires pre-fetching a number of results that is greater than the viewport can host (so that the scrollbar is always there). So we have to set a large static value that will work against all browsers and window sizes.

On smaller screens/windows, we end up pre-fetching far more rows than the client will need, and it only worsens over time. Indeed, a `(page)` event fires once we get past the number of rows that a single viewport can host, which is less than the `limit` set for external paging (it wasn't done in this example, but it makes no difference). I.e. the `pageSize` that triggers `(page)` event is [calculated ](https://github.com/swimlane/ngx-datatable/blob/master/src/components/datatable.component.ts#L905)based on the height of the table / `rowHeight`. So we end up fetching more and more "background" pages over time, and the scrollbar position approaches the ratio of `pageSize` / `limit`.

An alternative to that would be to use the `(scroll)` event, something like
```
@Component({
  selector: 'my-app',
  template: `
    <ngx-datatable
      class='material'
      [rows]='rows'
      [columnMode]="'force'"
      [headerHeight]="headerHeight"
      [rowHeight]="rowHeight"
      [scrollbarV]="true"
      [loadingIndicator]="isLoading"
      (scroll)="onScroll($event.offsetY)"
    >
    ...
  `
})
export class App implements OnInit {
  
  readonly headerHeight = 50;
  readonly rowHeight = 50;
  
  rows: any[] = [];
  isLoading: boolean;
  
  constructor(private el: ElementRef) { }
  
  ngOnInit() {
    this.onScroll(0);
  }
  
  onScroll(offsetY: number) {
    const rowsHeight = this.el.nativeElement.getBoundingClientRect().height - this.headerHeight;
    if (!this.isLoading && offsetY + rowsHeight >= this.rows.length * this.rowHeight) {
      this.loadPage(this.lastPage, Math.ceil(rowsHeight / this.rowHeight));
    }
  }
  
  loadPage(pageIndex: integer, pageSize: integer) {
    this.isLoading = true;
    this.fetch(pageSize).then(results => {
      this.rows.push(...results);
      this.isLoading = false;
    });
  }
}
```
This approach ensures that the next "page" is loaded only once we reach the end of the current view, and the number of results on that "page" is equal to the number that would fit on the current view. Now, if the user resizes the view (either enlarging or shrinking), next time the `(scroll)` event will take into account the new height of the view. Also, if the user scrolled the current view prior to enlarging it, the (scroll) event fires again upon resize and will fetch new results accordingly. The only time it fails is upon initial page load, where if you enlarge it the `(scroll)` event would not fire, unless you try to scroll first. Also, if any row detail is expanded, it will fire earlier in the process but that doesn't hurt as long as the number of expanded rows is small.

Overall this works much nicer and is closest to the experience of 'infinite scrolling' that people are used to. There could be further improvements to handle `(scroll)` events better. But at least, this solution relies on the public API and doesn't use any internals.

**Comment by amcdnl** [comment]
@dinvlad - should we remove the other demo in favor of this one?

**Comment by dinvlad** [comment]
Yes, have just submitted #839

**Comment by Matmo10** [comment]
Ended up here after reading various threads asking about infinite scrolling (with data that comes from a server for each "page"). 

First, I think it would be very helpful to clearly explain somewhere the difference(s) between "Scrolling server-side" and "virtual server-side".  I'm sure these terms are totally obvious to some people, but it's (probably) unclear to some people, myself included. 

Second, the [virtual server side](https://swimlane.github.io/ngx-datatable/#virtual-paging) demo is currently broken (`Error: query at index -1 failed: Fenwick tree array not initialized.`), which makes it even harder to distinguish between the two. 

I would imagine my use case is fairly common: when I approach the bottom of the table, I'd like to get some event that would allow me to go manually fetch the next page and add the rows to my [rows] input.  Looking at the source for the virtual server side pagination example, it seems like thats what I want, but any usage of `[externalPaging]="true"` causes the same error that the demo app is experiencing. 

TL;DR - The demo / code for "virtual paging server side" is broken.

**Comment by dinvlad** [comment]
@Matmo10 sorry, that was a while ago (it worked then) and I'm not sure what the demo page is doing now. I've actually moved away from Angular altogether (to Vue/Vuetify), so not sure I can help any futher.

@amcdnl may probably provide some pointers. My use case is as follows:
1) _virtually_ scroll the table, which means that _only_ the rows in the current browser _view_ are _rendered_, to avoid an ever-growing list of HTML elements that the user sees only a portion of at any given moment;
2) capture an event when the view reaches the end of the table;
3) load more rows from the server and append them to the table; alternatively, load _all_ of 1..N rows (where N is the _new_ "length" of the table) and replace the table with these rows. The latter approach addresses the issue of inconsistent updates between paging requests (so you don't get duplicate/missing rows due to the fact that the data may have been modified on the server between the paging requests). This approach also maps naturally to server-side paging mechanisms in many search engines (if you use one instead of a database), where even if you request rows M..N it will still effectively query all 1..N rows, so there's no benefit of asking only for M..N other than response size.
4) In both cases, ngx-datatable should automatically re-render the virtual view at the current position.

**Comment by Matmo10** [comment]
Thanks for the info. Sounds like my use case is identical - maybe I should try some older versions.

I thought the `virtualization` @Input is what controlled whether the table utilized virtualization or not, though. I'm still not sure why one of the server side pagination examples is considered virtual(ized) while the other is not since neither change the `virtualization` @Input.
